File: core/src/main/java/com/feign/framework/retry/RetryPolicy.java
package com.feign.framework.retry;

/**
 * Abstraction for retry policies.
 *
 * @author Feign Framework Team
 * @version 1.0.0-SNAPSHOT
 */
public interface RetryPolicy {

    /**
     * Determines if a retry should be attempted for a given exception.
     *
     * @param e the exception that occurred
     * @param retryCount the current retry count (0 for the first attempt)
     * @return true if the request should be retried, false otherwise
     */
    boolean canRetry(Exception e, int retryCount);

    /**
     * Gets the maximum number of retry attempts.
     *
     * @return the maximum retry count (0 means no retries)
     */
    int getMaxRetries();

    /**
     * Gets the base interval between retry attempts in milliseconds.
     *
     * @return the base interval in milliseconds
     */
    long getRetryInterval();

    /**
     * Gets the specific load balancer type this retry policy applies to.
     *
     * @return the load balancer type, or null if it applies to all types
     */
    String getLoadBalancerType();

    /**
     * Checks if this retry policy is enabled.
     *
     * @return true if retries are enabled, false otherwise
     */
    boolean isEnabled();
}

File: python-impl/src/feign/retry/retry_policy.py
"""Base abstract class for retry policies"""
from abc import ABC, abstractmethod
from typing import Optional


class RetryPolicy(ABC):
    """Abstract base class for retry policies"""

    @abstractmethod
    def can_retry(self, retry_count: int, last_exception: Optional[Exception]) -> bool:
        """Determine if a retry should be attempted

        Args:
            retry_count: Current retry attempt count (0-based)
            last_exception: The last exception that occurred

        Returns:
            True if retry should be attempted, False otherwise
        """
        pass

    @abstractmethod
    def max_retries(self) -> int:
        """Get the maximum number of retries

        Returns:
            Maximum retry count
        """
        pass

    @abstractmethod
    def retry_interval(self) -> int:
        """Get the retry interval in milliseconds

        Returns:
            Retry interval in milliseconds
        """
        pass
